cpp/src/arrow/compute/kernels/take.h

```diff
@@ -0,0 +1,94 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#pragma once
+
+#include <memory>
+
+#include "arrow/compute/kernel.h"
+#include "arrow/status.h"
+#include "arrow/util/visibility.h"
+
+namespace arrow {
+
+class Array;
+
+namespace compute {
+
+class FunctionContext;
+
+struct ARROW_EXPORT TakeOptions {
+  enum OutOfBoundsBehavior {
```

Comment:
While I think `RAISE`, `UNSAFE` and even `TO_NULL` make sense to have, these affect `Take`'s behavior so fundamentally, that We should have different `API`s for them instead of passing as an option.

AFAICT the rest of the compute functions follow the `RAISE` behavior, so I think `Take` should do as well. 

To support `UNSAFE` / more performant alternatives I suggest to choose a convention We can apply across the whole `compute` module, for example having `UnsafeTake` (similarly like other functions are named in arrow) rather than passing it as an option.

Lastly We could express `TO_NULL` differently, like using `Take` in conjunction with another kernel function which fills the arrays with nulls.